### PR TITLE
Implement custom (OAuth) flow

### PR DESCRIPTION
**Pull request by rflprr**
- Implement auth in accordance with best practices
- Improve error handling
- Clean-up and re-organize code
- Fix lint issues (in accordance with chosen eslint config)

GitHub issues:
https://github.com/datadotworld/tableau-connector/issues/41
https://github.com/datadotworld/tableau-connector/issues/38
https://github.com/datadotworld/tableau-connector/issues/27

### CODE REVIEW

File: src/TableauConnector.js [hunk 1]
```diff
@@ -47,103 +49,100 @@ const schemaMap = {
...
 
-  getApiEndpoint = () => {
-    const datasetCreds = JSON.parse(tableau.connectionData)
-    axios.defaults.headers.common['Authorization'] = `Bearer ${datasetCreds.apiToken}`
-    return `https://query.data.world/${datasetCreds.queryType || 'sql'}/${datasetCreds.dataset}`
-  }
+    this.authenticate()
+      .then(accessToken => this.validateAccessIfNeeded(accessToken))
+      .then(accessToken => {
+        const hasAuth = !!accessToken && accessToken.length > 0
```
Comment (alevicki) [inline src/TableauConnector.js]: The length check is unnecessary as an empty string is falsey

File: src/TableauConnector.js [hunk 2]
```diff
@@ -304,17 +308,91 @@ export default class TableauConnector {
...
-      tableau.log('There was an error retrieving the data')
+      dataCallback()
+      utils.log('SUCCESS: Data')
+    }).catch(error => {
       tableau.abortWithError(error)
+      utils.log(`FAILURE: Data (${error})`)
+    })
+  }
+
+  redirect401 (req) {
```
Comment (alevicki) [inline src/TableauConnector.js]: You can accomplish this with an interceptor, https://github.com/axios/axios#interceptors
That does assume you always want to perform that action on a 401.

File: src/api.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+axios.defaults.headers['Accept'] = 'application/json'
+
+const runSql = (dataset, query) => {
+  return axios.post(
+    `${basePathQuery}/sql/${dataset}`,
+    queryString.stringify({query}),
+    {
+      headers: {
+        'authorization': 'Bearer ' + getApiKey(true),
```
Comment (alevicki) [inline src/api.js]: You can use a template literal here

File: src/api.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    })
+}
+
+const runSparql = (dataset, query) => {
+  return axios.post(
+    `${basePathQuery}/sparql/${dataset}`,
+    queryString.stringify({query}),
+    {
+      headers: {
+        'authorization': 'Bearer ' + getApiKey(true),
```
Comment (alevicki) [inline src/api.js]: You can use a template literal here

File: src/api.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  }
+  return apiFn(dataset, query)
+}
+
+const getUser = () => {
+  return axios.get(
+    `${basePath}/user`,
+    {
+      headers: {
+        'authorization': 'Bearer ' + getApiKey(true)
```
Comment (alevicki) [inline src/api.js]: You can use a template literal here

File: src/api.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import axios from 'axios'
+import * as queryString from 'query-string'
+import { getApiKey } from './auth'
+
+const basePath = 'https://api.data.world/v0'
+const basePathQuery = 'https://query.data.world'
+
+axios.defaults.headers['Accept'] = 'application/json'
+
+const runSql = (dataset, query) => {
```
Comment (alevicki) [inline src/api.js]: `runSql` and `runSparql` appear to be identical except for the url (sql vs sparql).  They could be reduced to a single function that takes the type of query, or moved into the `runQuery` function.

File: src/TableauConnector.js (hunk 2 above)
Comment (rflprr) [reply src/TableauConnector.js]: This form of error handling only applies in some cases. `TableauConnector.js` knows when to use it and it's not aware of `axios`.

File: src/TableauConnector.js [hunk 7]
```diff
@@ -17,12 +17,14 @@
  * data.world, Inc. (http://data.world/).
  */
 
-import axios from 'axios'
-import queryString from 'query-string'
+import * as auth from './auth'
```
[the remark this answers is not in the document]
Comment (alevicki) [reply src/TableauConnector.js]: Not in bulk like this does.  You can alias each individual function, but that isn't really the same thing.

This is mostly a style and code reading preference so treat the initial comment as such.

**Pull request merged**